src/Features/Core/Portable/Diagnostics/DiagnosticService.cs

```diff
@@ -314,17 +322,35 @@ private void AssertIfNull(ImmutableArray<DiagnosticData> diagnostics)
         }
 
         [Conditional("DEBUG")]
-        private void AssertIfNull(DiagnosticData diagnostic)
+        private void AssertIfNull<T>(T obj) where T : class
         {
-            if (diagnostic == null)
+            if (obj == null)
             {
                 Contract.Requires(false, "who returns invalid data?");
             }
         }
 
+        [Conditional("DEBUG")]
+        public void AssertCleanup()
+        {
+            foreach (var idAndData in _map.Values.ToList())
+            {
+                foreach (var data in idAndData.Values.ToList())
+                {
+                    AssertIfNull(data.Workspace);
+                }
+            }
+        }
+
         private struct Data : IEquatable<Data>
         {
-            public readonly Workspace Workspace;
+            private static readonly ConditionalWeakTable<Workspace, WeakReference<Workspace>> s_cache =
```

Comment:
CWT is basically pool. why create so many redundant weak reference objects? if there are 300 analyzers (basically install style cop analyzers), worst case will be 300 x number of projects. which pointing to same workspace. why just use 1 rather than creating all those weak reference?